### PR TITLE
AESA Radar

### PR DESCRIPTION
**Pull request by jjosborne-dstl**
Added an AESA (Active electronically scanned array) radar sensor along with functions for the movement of the radar beam and the beam shape. The radar sensor calculates the probability of detection of a state and generates a measurement based on that probability.

### CODE REVIEW

File: stonesoup/sensor/radar/radar.py [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+                           doc="Is the Swerling 1 case used. If True the RCS"
+                               " of the target will change for each timestep. "
+                               "The random RCS follows the probability "
+                               "distribution of the Swerling 1 case.")
+    rcs = Property(float, default=None, doc="The radar cross section of "
+                                            "targets in meters squared.")
+
+    probability_false_alarm = Property(Probability, default=1e-6,
+                                       doc="Probability of false alarm used in"
+                                           " the North's approximation")
```
Comment (sdhiscocks) [inline stonesoup/sensor/radar/radar.py]: The above arguments could be made neater by laying out like this:
```python
rotation_offset = Property(
    StateVector, default=StateVector([0, 0, 0]),
    doc="A 3x1 array of angles (rad), specifying "
        "the radar orientation in terms of the "
        "counter-clockwise rotation around the "
        ":math:`x,y,z` axis. i.e Roll, Pitch and Yaw.")
translation_offset = Property(
    StateVector, default=StateVector([0, 0, 0]),
    doc="The radar position in 3D Cartesian space.[x,y,z]")
mapping = Property(
    np.array, default=[0, 1, 2],
    doc="Mapping between or positions and state "
        "dimensions. [x,y,z]")
...
```

File: stonesoup/sensor/radar/radar.py [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+        noise_figure = 10 ** (self.receiver_noise / 10)
+        loss = 10 ** (self.loss / 10)
+        # calculate part of snr that is independent of:
+        #   rcs, transmitted power, gain and range
+        self.snr_constant = (const.c ** 2 * self.number_pulses *
+                             self.duty_cycle) / (64 * np.pi ** 3 * const.k *
+                                                 temp * self.band_width *
+                                                 noise_figure * self.frequency
+                                                 ** 2 * loss)
+        # calculate range resolution for compressed pulse
```
Comment (sdhiscocks) [inline stonesoup/sensor/radar/radar.py]: Not sure this should be here?

File: stonesoup/sensor/radar/radar.py [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        temp = 290  # noise reference temperature (room temperature kelvin)
+        # convert from dB
+        noise_figure = 10 ** (self.receiver_noise / 10)
+        loss = 10 ** (self.loss / 10)
+        # calculate part of snr that is independent of:
+        #   rcs, transmitted power, gain and range
+        self.snr_constant = (const.c ** 2 * self.number_pulses *
```
Comment (sdhiscocks) [inline stonesoup/sensor/radar/radar.py]: This could be made a calculated property.

File: stonesoup/sensor/radar/radar.py [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+                             self.duty_cycle) / (64 * np.pi ** 3 * const.k *
+                                                 temp * self.band_width *
+                                                 noise_figure * self.frequency
+                                                 ** 2 * loss)
+        # calculate range resolution for compressed pulse
+
+    def swerling_1(self, rcs):
```
Comment (sdhiscocks) [inline stonesoup/sensor/radar/radar.py]: Should this be a "private" method?

File: stonesoup/sensor/radar/radar.py [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+        sky_state : The target state.
+
+        Returns
+        -------
+        det_prob : Probability of detection.
+        snr : Signal to noise ratio.
+        rcs : RCS after the Swerling 1 case is applied.
+        directed_power : Power in the direction of the target.
+        spoiled_gain : Gain in decibels with beam spoiling applied.
+        spoiled_width : Beam width with beam spoiling applied.
```
Comment (sdhiscocks) [inline stonesoup/sensor/radar/radar.py]: Syntax of these is wrong, as after colon should be type, and description on line below. See [5. Returns](https://numpydoc.readthedocs.io/en/latest/format.html#sections).

File: stonesoup/sensor/radar/radar.py [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+        """
+        if hasattr(sky_state, 'rcs') and sky_state.rcs is not None:
```
Comment (sdhiscocks) [inline stonesoup/sensor/radar/radar.py]:
```python
if getattr(sky_state, 'rcs', None) is not None:
    ...
```

File: stonesoup/sensor/radar/radar.py [hunk 7]
```diff
@@ -0,0 +1,453 @@
...
+            # use state's rcs if it has one
+            rcs = sky_state.rcs
+        else:
+            rcs = self.rcs
+        # apply swerling 1 case?
+        if self.swerling_on:
+            rcs = self.swerling_1(rcs)
```
Comment (sdhiscocks) [inline stonesoup/sensor/radar/radar.py]: Is the `swerling_1()` method needed if this is the only place it is used?

**Comment by sdhiscocks** [comment]
For anyone else reviewing this, it's easier to see changes via command below rather than using GitHub interface due to filename changes:
```
git diff -M25% f9a893385fff06e5f45efcaa30d109728c96696d
```

**Comment by etfrogers-dstl** [comment]
I have been working on this with Jonathan, and as part of the follow-on work realized that this branch had diverged from master and will not merge cleanly. 

I have therefore (for my own work) created another branch which contains the changes in the develop_radar branch, but rebased onto master (as of Friday afternoon), which should merge a lot more easily. As this was a bit painful, and I wouldn't want anyone else to repeat it, is there a way of switching the pull request to the new branch, rather than this one? Or possibly we should close this one and open a new one?

**Comment by sdhiscocks** [comment]
@etfrogers-dstl: Thanks. Probably best to open a new PR and close this one.

**Comment by etfrogers-dstl** [comment]
This pull request is replaced by #159 (which should contain the same code in a more easy-to-merge form.)

This request is now being closed

**Pull request closed without merging**